common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use libra_logger::prelude::*;
+use std::{collections::HashMap, fmt::Debug, hash::Hash, time::Instant};
+use tokio::time::Duration;
+
+/// Interface for a keyed rate limiter.  This should allow us to
+/// drop in a different keyed rate limiter if required.
+pub trait KeyedRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    /// Determine if a throttle is needed on the key for the given count
+    /// This does not block, or wait and returns `true` if throttled
+    fn throttle(&mut self, key: Key, count: u64) -> bool;
```

Comment:
is `bool` the right return value here? if I'm throttled, presumably I need either some wait time (so I myself can wait until the next time where I might be able to get a permit) or I need to register my task's Waker with the rate limiter so it can wake me up when it's my turn.

Reply:
I'm leaving it up to the requesting service to handle it in this first iteration, there's a TODO in there for something like this.

Here's our problem, our downstream services don't retry, but since we currently don't apply backpressure we'd have to just wait until it's my turn.  But, what happens when you've got a ton of requests coming in waiting, how do we load shed so we don't just keep accepting requests that then get immediately thrown into waiting.

I have a feeling that different situations would prescribe different solutions, including how do you then provide fairness on the waking based on order, size, and key.  So, for now I *want* to just return that it's been throttled, and initially drop the request.

Reply:
Gotcha, maybe we return (like governor), `Result<(), Duration>` to signal "hey you're throttled and here's a hint on when you can probably next get a permit". Then callers can decide to either wait (if they're applying backpressure) or load shed o/w.

Reply:
Yeah, that's a little different, where the rate limiter handles the jitter and the delay in which you'll be sending your next request, which is interesting.  I'll take a look into it and see if it'll give us some results.

I highly expect this piece to evolve over time :)